packages/ActionBar/src/hooks/useFilter/useFilter.js

```diff
@@ -0,0 +1,112 @@
+/* eslint-disable no-param-reassign */
+import React from "react";
+import nanoid from "nanoid";
+import produce from "immer";
+import Filter from "../../components/Filter";
```

Comment:
That is weird that the hook depends on the component. Maybe common dependencies, such as `defaultRulesByType`, can be extracted?

Reply:
@nahumzs @allison-c Also, I'm not sure why we separate hooks from components. We need to separate concerns, but not the form. Here, `Filter` and `useFilter` are logically connected. Imagine we had no hooks in React. Then, instead of `useFilter` you'd create another component, something like `FilterConsumer`, and would put in next to `Filter`. Then why we need a different treatment for `useFilter`?

Reply:
The same goes for `useSort` and `useColumnsArrangement`

Reply:
My idea is to let consumers can use `Filter` as a controlled component but also an uncontrolled component if they use `useFilter`, same as `Sort` and `ColumnArrangement`. 
Maybe provide a `FilterController` component and a `Filter` component is a better option and easier to understand? 
We also take charge of the data filtering part in `useFilter` hook, so even though if we have a controlled filter and uncontrolled filter, we'll still need a `useFilter` hook to help the data filtering

Reply:
Yeah, the idea was to use the hooks if you want to convert them into an uncontrolled components, that's the idea behind hook to reuse logic, I think its cleaner than creating another component that might only confuse our users.

Reply:
What I'm not liking here is, at least from seeing the stories, I don't really care what this hook returns. E.g.
```jsx
  const {
    orderedColumns,
    handleChangeVisibility,
    handleShowAll,
    handleHideAll,
    handleChangeOrder,
  } = useColumnsArragment(defaultColumnsForArrangement);

  return (
    <ColumnsArrangement
      columns={orderedColumns}
      onChangeOrder={handleChangeOrder}
      onHideAll={handleHideAll}
      onShowAll={handleShowAll}
      onChangeVisibility={handleChangeVisibility}
    />
```
All those variables only exist to immediately pass them on to the component, I don't need them otherwise. This could be made slightly better
```jsx
  const whatever = useColumnsArragment(defaultColumnsForArrangement);

  return (
    <ColumnsArrangement {...whatever} />
  );
```
but I'm not sure that's too much better than
```jsx
return <ColumnsArrangement.Default columns={defaultColumns} />;
```
?

Reply:
```js
const whatever = useColumnsArragment(defaultColumnsForArrangement);
  return (
    <ColumnsArrangement {...whatever} />
  );
```

In that regard you are correct I don't know why I didn't just spread the Hook props on the component brain fart (maybe the pressure to delivering the demo for testing purposes). 

But I do think using hooks to abstract the logic it's the appropriate way to go for these cases, at the end, it's the main reason why they were created to be able to reuse logic. 

I recently read the updated version of the `reducer pattern` by Kent Dodds where he moves away from a render function to hooks and it is expecting it to use it in his downshift library. https://kentcdodds.com/blog/the-state-reducer-pattern-with-react-hooks

^ So that gives me confidence that the approach of using hooks it's the correct way to go.